<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Manual</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">DBSLMM</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
<li>
  <a href="Manual.html">Manual</a>
</li>
<li>
  <a href="Validate.html">External Validate</a>
</li>
<li>
  <a href="Scripts.html">Scripts for Other Methods</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Manual</h1>

</div>


<div id="overview" class="section level1">
<h1>Overview</h1>
<p>We upload two versions of DBSLMM, including default version (p-value threshold=1e-06 and LD threshold=0.1) and tuning version. Specially, tuning version needs a validation data, like LDpred, lassosum adn C+T.</p>
</div>
<div id="dependencies" class="section level1">
<h1>Dependencies</h1>
<ul>
<li>R <br> We use <a href="https://www.r-project.org/">R program</a> to wrap <a href="https://www.cog-genomics.org/plink/1.9/">PLINK</a> and executable file <code>dbslmm</code>. We use two R pakages <code>optparse</code> and <code>data.table</code>. You can install them by: <br></li>
</ul>
<pre class="r"><code>install.packages(c(&quot;data.table&quot;, &quot;optparse&quot;), dependencies=TRUE)</code></pre>
<ul>
<li>PLINK <br> DBSLMM treats the large effect SNPs as fixed effect and treats small effect SNPs as random effects. In the paper, we use the pruning and clumping strategy to select large effect SNPs.</li>
</ul>
</div>
<div id="example-data" class="section level1">
<h1>Example data</h1>
<ul>
<li>Summary statistics (<code>summary_gemma_chr1.assoc.txt</code>) <br> The summary statistics is estiamted from 2000 samples from the UKB data. SNPs are the first 100,000 SNPs from chromosome 1. <br></li>
<li>Reference data (<code>ref_chr1.bed</code>, <code>ref_chr1.bim</code> and <code>ref_chr1.fam</code>) <br> We sampled 300 samples from the 1000 Gemone Project data. <br></li>
<li>Test data (<code>test_chr1.bed</code>, <code>test_chr1.bim</code> and <code>test_chr1.fam</code>) <br> We sampled 103 samples from the 1000 Gemone Project data. <br> Note: The phenotype of test data is simulated by normal distribution. Therefore, you can not use the test data to evaluate the prediction performance.</li>
</ul>
</div>
<div id="input-file-format" class="section level1">
<h1>Input file format</h1>
<ul>
<li>Summary statistics (<a href="https://github.com/genetics-statistics/GEMMA">GEMMA format</a>) <br></li>
</ul>
<pre class="bash"><code>chr rs  ps  n_mis   n_obs   allele1 allele0 af  beta    se  p_wald
1   rs58276399  731718  99  2301    C   T   0.110   1.272675e-01    3.461097e-02    2.410969e-04
1   rs141242758 734349  98  2302    C   T   0.109   1.276907e-01    3.463008e-02    2.317011e-04
1   rs28544273  751343  34  2366    A   T   0.120   1.320775e-01    3.276319e-02    5.720490e-05
1   rs28527770  751756  31  2369    C   T   0.121   1.364192e-01    3.269947e-02    3.128107e-05
1   rs3115860   753405  16  2384    C   A   0.128   1.470137e-01    3.158378e-02    3.421012e-06
1   rs529912679 753430  17  2383    C   T   0.125   1.492779e-01    3.201234e-02    3.285182e-06
1   rs2073813   753541  22  2378    A   G   0.126   1.422969e-01    3.182372e-02    8.132450e-06
1   rs3131969   754182  16  2384    A   G   0.129   1.487274e-01    3.155196e-02    2.571517e-06
1   rs3131968   754192  16  2384    A   G   0.129   1.487274e-01    3.155196e-02    2.571517e-06
1   rs3131967   754334  16  2384    T   C   0.129   1.487274e-01    3.155196e-02    2.571517e-06
1   rs3115858   755890  3   2397    A   T   0.129   1.467534e-01    3.146006e-02    3.259725e-06
1   rs181250764 756533  47  2353    T   C   0.119   1.534308e-01    3.305476e-02    3.641537e-06
1   rs3131962   756604  0   2400    A   G   0.130   1.464455e-01    3.141295e-02    3.304095e-06
1   rs6699990   756912  0   2400    A   G   0.130   1.464455e-01    3.141295e-02    3.304095e-06
1   rs3115853   757640  3   2397    G   A   0.130   1.487149e-01    3.138661e-02    2.282682e-06
1   rs4951929   757734  1   2399    C   T   0.129   1.469647e-01    3.142780e-02    3.083623e-06
1   rs4951862   757936  1   2399    C   A   0.129   1.469647e-01    3.142780e-02    3.083623e-06
1   rs3131956   758144  1   2399    A   G   0.129   1.469647e-01    3.142780e-02    3.083623e-06
1   rs3131954   758626  3   2397    C   T   0.129   1.489328e-01    3.145579e-02    2.321970e-06</code></pre>
<ul>
<li>Reference panel (<a href="https://www.cog-genomics.org/plink/1.9/formats">PLINK format</a>) <br> The same file name of bed, bim and fam files.</li>
<li>Block information <br> We use the block information from <a href="https://academic.oup.com/bioinformatics/article/32/2/283/1743626/Approximately-independent-linkage-disequilibrium">Berisa and Pickrell (2015)</a>. You can download the block information from <a href="https://bitbucket.org/nygcresearch/ldetect-data/src/master/" class="uri">https://bitbucket.org/nygcresearch/ldetect-data/src/master/</a>.</li>
</ul>
</div>
<div id="input-parameters" class="section level1">
<h1>Input parameters</h1>
<div id="dbslmm" class="section level2">
<h2>DBSLMM</h2>
<p>We prepare all the input file in the folder <code>test_dat</code>. You can output the explanation for each parameter:</p>
<pre class="bash"><code>DSBLMM=/your/path/DBSLMM/software/DBSLMM.R
Rscript ${DBSLMM} -h
Rscript ${DBSLMM} --help</code></pre>
<p>The details is:</p>
<pre class="bash"><code>--summary=CHARACTER
        INPUT: the summary statistics (gemma output format)
--plink=CHARACTER
        INPUT: the perfix of Plink software
--dbslmm=CHARACTER
        INPUT: the perfix of dbslmm software
--ref=CHARACTER
        INPUT: the perfix of reference panel
--block=CHARACTER
    INPUT: the block information (Berisa and Pickrell 2015)
--outPath=CHARACTER
        INPUT: the output path
--n=CHARACTER
        INPUT: the sample size of summary data
--nsnp=CHARACTER
        INPUT: the number of SNPs in whole genome
--h2=CHARACTER
        INPUT: the heritability of trait
--type=CHARACTER
        INPUT: type of DBSLMM (default: default version)
--r2=CHARACTER
        INPUT: the cutoff of SNPs clumping (default: 0.1)
--pv=CHARACTER
        INPUT: the cutoff of SNPs pruning (default: 1e-6)
--mafMax=CHARACTER
        INPUT: the maximium of the difference between reference panel and summary data (default:0.2)
--thread=CHARACTER
        INPUT: the number of threads (default: 5)</code></pre>
</div>
<div id="tune" class="section level2">
<h2>TUNE</h2>
<p>You can output the explanation for each parameter:</p>
<pre class="bash"><code>TUNE=/your/path/DBSLMM/software/TUNE.R
Rscript ${TUNE} -h
Rscript ${TUNE} --help</code></pre>
<p>The detail is:</p>
<pre class="bash"><code>--phenoPred=CHARACTER
        INPUT: the predicted phenotype (PLINK output)
--phenoVal=CHARACTER
        INPUT: the validation phenotype file and column number
--index=CHARACTER
        INPUT: four different indexes (MSE and R2 for continuous trait, Brier score and AUC for binary trait)
--pvRange=CHARACTER
        INPUT: the range of p-value threshold
--ldRange=CHARACTER
        INPUT: the range of LD threshold
--cov=CHARACTER
        INPUT: covariate
--chr=CHARACTER
        INPUT: chromosome</code></pre>
</div>
</div>
<div id="output-file-format" class="section level1">
<h1>Output file format</h1>
<p>The example of output file is:</p>
<pre class="bash"><code>rs13302957 G 0.546652 1.82642 1
rs3748588 T 0.273432 1.64563 1
rs74045047 A -0.282306 -0.5378 1
rs3845292 G 0.159231 0.22554 1
rs113288277 T -0.0524222 -0.18273 1
rs112797925 A 0.252739 1.12303 1
rs12743678 A 0.741517 1.51564 1
rs141242758 C 0.00548007 0.0124342 0
rs28544273 A 0.0114189 0.0248473 0
rs3115860 C 0.00628248 0.013297 0
rs2073813 A 0.00771072 0.0164301 0
rs3131969 A 0.00191854 0.00404717 0
rs3131968 A 0.00191854 0.00404717 0
rs3131967 T 0.00191854 0.00404717 0
rs3115858 A 0.00415565 0.00876637 0
rs3131962 A 0.0112978 0.0237545 0
rs3115853 G 0.00251826 0.00529485 0
rs4951929 C 0.00443819 0.0093624 0
rs4951862 C 0.00256037 0.00540112 0
rs3131956 A 0.00341574 0.00720553 0</code></pre>
<p>The first column is SNP ID. The second column is allele code. The third code is scaled effect size. The forth is non-scaled effect size (using MAF from summary statistics). You can also use the MAF from other reference panel to estimate effect size. The fifth column is the index of whether it is large effect or not (1: large effect, 0: small effect). This output format can be directly used to <code>score</code> function of PLINK.</p>
</div>
<div id="example-code" class="section level1">
<h1>Example code</h1>
<div id="parameters-initialization" class="section level2">
<h2>Parameters initialization</h2>
<pre class="bash"><code>### change file permission for dbslmm
dbslmm=/your/path/DBSLMM/software/dbslmm
chmod 777 ${dbslmm}
### Parameters for DBSLMM
mkdir /your/path/DBSLMM/test_dat/out
let chr=1
DBSLMM=/your/path/DBSLMM/software/DBSLMM.R
summf=/your/path/DBSLMM/test_dat/summary_gemma_chr
outPath=/your/path/DBSLMM/test_dat/out/
plink=/your/path/plink-1.9
ref=/your/path/DBSLMM/test_dat/ref_chr
blockf=/your/path/DBSLMM/test_dat/chr
m=`cat ${summf}${chr}.assoc.txt | wc -l` 
h2=0.5
nobs=`sed -n &quot;2p&quot; ${summf}${chr}.assoc.txt | awk &#39;{print $5}&#39;`
nmis=`sed -n &quot;2p&quot; ${summf}${chr}.assoc.txt | awk &#39;{print $4}&#39;`
n=$(echo &quot;${nobs}+${nmis}&quot; | bc -l)</code></pre>
<p>Note: In real data, <code>m</code> is the snp number for whole genome and <code>h2</code> is esitmated by LDSC(<a href="https://github.com/bulik/ldsc" class="uri">https://github.com/bulik/ldsc</a>) or GEMMA(<a href="https://github.com/genetics-statistics/GEMMA" class="uri">https://github.com/genetics-statistics/GEMMA</a>). For the binary traits, we recommend to use liability scale beritability.</p>
</div>
<div id="default-version" class="section level2">
<h2>Default version</h2>
<pre class="bash"><code>### DBSLMM
Rscript ${DBSLMM} --summary ${summf}${chr}.assoc.txt --outPath ${outPath} --plink ${plink} --dbslmm ${dbslmm} --ref ${ref}${chr} --n ${n} --nsnp ${m} --block ${blockf}${chr}.bed --h2 0.5
### Predict
bfilete=/your/path/test_dat/test_chr
est=/your/path/test_dat/out/summary_gemma_chr
InterPred=/your/path/out/internal_pred_chr
## plink 1.9
plink=/your/path/plink1.9
${plink} --bfile ${bfilete}${chr} --score ${est}${chr}.assoc.dbslmm.txt 1 2 4 sum --out ${InterPred}${chr}
rm ${InterPred}${chr}.log
## plink 2
plink=/your/path/plink2
${plink} --bfile ${bfilete} --score ${est}${chr}.assoc.dbslmm.txt 1 2 4 cols=+scoresums --out ${InterPred}${chr}
rm ${InterPred}.log</code></pre>
<p>Note: Please check the reference panel. DBSLMM is not accept genotype missing for all samples in reference panel.</p>
</div>
<div id="tuning-version" class="section level2">
<h2>Tuning version</h2>
<div id="fit-model-with-different-p-value-and-ld-threshold" class="section level3">
<h3>Fit model with different p-value and LD threshold</h3>
<pre class="bash"><code>### DBSLMM for chromosome 1
for pv in 1e-05 1e-06 1e-07 1e-08
do
for r2 in 0.05 0.1 0.15 0.2 0.25
do
Rscript ${DBSLMM} --summary ${summf}${chr}.assoc.txt --outPath ${outPath} --plink ${plink} --dbslmm ${dbslmm} --ref ${ref}${chr} --pv ${pv} --r2 ${r2} --mafMax 1 --n ${n} --nsnp ${m} --type t --block ${blockf}${chr}.bed --h2 ${herit}
val=/your/path/DBSLMM/test_dat/val_chr
InterPred=/your/path/DBSLMM/test_dat/out/pheno_gemma_chr${chr}_pv${pv}_r${r2}
plink-1.9 --bfile ${val}${chr} --score ${outPath}summary_gemma_chr${chr}_pv${pv}_r${r2}.dbslmm.txt 1 2 4 sum --silent --out ${InterPred}
rm ${outPath}summary_gemma_chr${chr}_pv${pv}_r${r2}.dbslmm.badsnps
rm ${InterPred}.log
if [ -f &quot;${InterPred}.nopred&quot; ];then
rm ${InterPred}.nopred
fi
done
done</code></pre>
<p>Note: You can submit many jobs for different p-value thresholds, LD thresholds and chromosomes.</p>
</div>
<div id="select-the-best-parameter-combination" class="section level3">
<h3>Select the best parameter combination</h3>
<pre class="bash"><code>tuningPred=/your/path/DBSLMM/script/TUNE.R
InterPred=/your/path/DBSLMM/test_dat/out/pheno_gemma_chr
phenoVal=/your/path/DBSLMM/test_dat/val_chr
index=r2 #mse, auc or bs
## chromosome 1 
Rscript ${TUNE} --phenoPred ${InterPred} --phenoVal ${phenoVal}${chr}.fam,6 --pvRange 1e-05,1e-06,1e-07,1e-08 --ldRange 0.05,0.1,0.15,0.2,0.25 --index ${index} --chr 1
rbest=`cat ${InterPred}_rbest.${index}`
pbest=`cat ${InterPred}_pbest.${index}`
mv ${outPath}summary_gemma_chr${chr}_pv${pbest}_r${rbest}.dbslmm.txt ${outPath}summary_gemma_chr${chr}_best.dbslmm.txt rm ${outPath}summary_gemma_chr${chr}_pv*
rm ${InterPred}*.profile

## whole genome
Rscript ${TUNE} --phenoPred ${InterPred} --phenoVal ${phenoVal}${chr}.fam,6 --pvRange 1e-05,1e-06,1e-07,1e-08 --ldRange 0.05,0.1,0.15,0.2,0.25 --index ${index} 
rbest=`cat ${InterPred}_rbest.${index}`
pbest=`cat ${InterPred}_pbest.${index}`
for chr in `seq 1 22`
do
mv ${outPath}summary_gemma_chr${chr}_pv${pbest}_r${rbest}.dbslmm.txt ${outPath}summary_gemma_chr${chr}_best.dbslmm.txt rm ${outPath}summary_gemma_chr${chr}_pv*
done
rm ${InterPred}*.profile</code></pre>
</div>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
